src/TroykaTextLCD.cpp

```diff
@@ -1,173 +1,179 @@
-    }
-    if (charSize == 1) {
-        _displayFunction |= LCD_5x8DOTS;
-    } else if ((charSize == 2) && (_rows == 2)) {
-        _displayFunction |= LCD_5x10DOTS;
-    }
-    setNormalFunction();
-    setExtendFunction();
-    command(LCD_EX_SETBIASOSC | LCD_BIAS_1_5 | LCD_OSC_183HZ);
-    command(LCD_EX_FOLLOWERCONTROL | LCD_FOLLOWER_ON | LCD_RAB_2_00);
-    delay(200);
-    setNormalFunction();
-
-    _displayControl = 0x00;
-    _displayControl = LCD_DISPLAYON | LCD_CURSOROFF | LCD_BLINKOFF;
-    command(LCD_DISPLAYCONTROL | _displayControl);
-    
-    clear();
-
-    _displayMode = 0x00;
-    _displayMode = LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT;
-    command(LCD_ENTRYMODESET | _displayMode);
+  _cols = cols;
+  _rows = rows;
+  _charSize = charSize;
+  pinMode(_pinBacklight, OUTPUT);
+  setBrightness(255);
+
+  _wire->begin();
+  delay(40);
+
+  _displayFunction = LCD_8BITMODE;
+  if (_rows == 1) {
+    _displayFunction |= LCD_1LINE;
+  } else if (_rows == 2) {
+    _displayFunction |= LCD_2LINE;
+  }
+  if (charSize == 1) {
+    _displayFunction |= LCD_5x8DOTS;
+  } else if ((charSize == 2) && (_rows == 2)) {
+    _displayFunction |= LCD_5x10DOTS;
+  }
+  setNormalFunction();
+  setExtendFunction();
+  command(LCD_EX_SETBIASOSC | LCD_BIAS_1_5 | LCD_OSC_183HZ);
+  command(LCD_EX_FOLLOWERCONTROL | LCD_FOLLOWER_ON | LCD_RAB_2_00);
+  delay(200);
+  setNormalFunction();
+
+  _displayControl = 0x00;
+  _displayControl = LCD_DISPLAYON | LCD_CURSOROFF | LCD_BLINKOFF;
+  command(LCD_DISPLAYCONTROL | _displayControl);
+
+  clear();
+
+  _displayMode = 0x00;
+  _displayMode = LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT;
+  command(LCD_ENTRYMODESET | _displayMode);
 }
 
 void TroykaTextLCD::setContrast(uint8_t contrast) {
-    setExtendFunction();
-    command(LCD_EX_CONTRASTSETL | (contrast & 0x0f));
-    command(LCD_EX_POWICONCONTRASTH | LCD_ICON_ON | LCD_BOOST_ON | ((contrast >> 4) & 0x03));
-    setNormalFunction();
-}
-
-void TroykaTextLCD::setBrightness(uint8_t brightness) {
+  setExtendFunction();
+  command(LCD_EX_CONTRASTSETL | (contrast & 0x0f));
+  command(LCD_EX_POWICONCONTRASTH | LCD_ICON_ON | LCD_BOOST_ON |
+          ((contrast >> 4) & 0x03));
+  setNormalFunction();
+}
+
+void TroykaTextLCD::setBrightness(uint16_t brightness) {
+  // Default PWM resolution 10 bit for board ESP8266
+  // Default PWM resolution 8 bit other boards
+  #if defined(ESP8266)
+    analogWrite(_pinBacklight, brightness << 2);
+  #else
     analogWrite(_pinBacklight, brightness);
+  #endif
 }
 
 void TroykaTextLCD::setIcon(uint8_t addr, uint8_t bit) {
-    setExtendFunction();
-    command(LCD_EX_SETICONRAMADDR | (addr & 0x0f));
-    write(bit);
-    setNormalFunction();
+  setExtendFunction();
+  command(LCD_EX_SETICONRAMADDR | (addr & 0x0f));
+  write(bit);
+  setNormalFunction();
 }
 
 void TroykaTextLCD::clear() {
-    command(LCD_CLEARDISPLAY);
-    delayMicroseconds(2000);
+  command(LCD_CLEARDISPLAY);
+  delayMicroseconds(2000);
```

Comment:
can you still reduce the delay? If I clear display in loop, I see the black display.

Reply:
The documentation for controller Display «ST7032» write, that the delay after clearing the screen should be more than 1.08 ms.